subscriptions/providers/apple_in_app/__init__.py

```diff
@@ -102,99 +110,167 @@ def check_payments(self, payments: Iterable[SubscriptionPayment]):
                 raise AppleSubscriptionNotCompletedError(payment.provider_transaction_id)
 
     @classmethod
-    def _get_validated_in_app_product(cls, response: AppleVerifyReceiptResponse) -> AppleInApp:
+    def _raise_if_invalid(cls, response: AppleVerifyReceiptResponse) -> None:
         if not response.is_valid or response.receipt.bundle_id != cls.bundle_id:
             raise AppleReceiptValidationError()
-        return one(response.receipt.in_apps)
 
-    @transaction.atomic
-    def _handle_receipt(self, request: Request, payload: AppleReceiptRequest) -> Response:
-        receipt = payload.transaction_receipt
+    def _get_plan_for_product_id(self, product_id: str) -> Plan:
+        search_kwargs = {
+            f'metadata__{self.codename}': product_id,
+        }
+        return Plan.objects.get(**search_kwargs)
 
-        # Validate the receipt. Fetch the status and product.
-        receipt_data = self.api.fetch_receipt_data(receipt)
-        single_in_app = self._get_validated_in_app_product(receipt_data)
+    def _handle_single_receipt_info(self,
+                                    user: User,
+                                    receipt_info: AppleLatestReceiptInfo) -> Optional[SubscriptionPayment]:
+        if receipt_info.cancellation_date is not None:
+            logger.error('Found a cancellation date in receipt: %s', receipt_info)
+            return None
```

Comment:
It's an error but you do nothing? Can you explain this place please?

Reply:
Hmm, I'm just refusing to handle this for now. This requires quite a few steps and is hard to test in any environment. I'd prefer to leave an error message with information helping further implementation the first time it happens than to handle it incorrectly.

What I'm aiming for here is to avoid handling it, the user gets the refund from Apple anyway and we keep the reminder of his subscription intact.

I've implemented handling of a refund, reduced the log to warning and added an additional explanation. Hope it helps.